---
title: "Population_Sim_Trial"
author: "Brandie QC"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Population Simulation Attempt

Suggested parameters from Julin:

-   1000 F2s (or just 1000 individuals)

    -   Done with runMacs

-   100 loci per trait with some kind of (expotnetial?) decay in effect
    size

    -   Use nQtlPerChr??

-   Traits:

    -   Germination

    -   Initial size

    -   Establishment

    -   Growth rate

    -   Flowering probability

        -   Ideally would depend on size... - do this through
            correlation matrix?

    -   Fruit per plant

        -   Ideally would depend on size...

    -   others?

Rishav demoed AlphaRSim:
<https://cran.r-project.org/web/packages/AlphaSimR/index.html>

-   Rishav's demo is on his GitHub:
    <https://github.com/rishavray/IPM_practice/blob/main/IPM_genetic_simulation.Rmd>

## Libraries

```{r}
library(AlphaSimR)
library(tidyverse)
logit <- brms::logit_scaled
inv_logit <- brms::inv_logit_scaled
```

## Create Founder Population

```{r}
founderPop = runMacs(
  nInd = 1000, #number of individuals to simulate 
  nChr = 10, #number of chromosomes to simulate - may not be important to match to Streps' 14
  segSites = NULL, #number of segregating sites to keep per chromosome.
  species = "GENERIC" #The GENERIC history is meant to be a reasonable all-purpose choice. It runs quickly and models a population with an effective populations size that has gone through several historic bottlenecks.
)
```

## Define Genetic Architecture for Traits

```{r}
SP <- SimParam$new(founderPop)
#Starts the process of building a new simulation by creating a new SimParam object and assigning a founder population to the class.

traitMeans <- c(germination.logit = logit(0.33), #logit to scale it (from gene --> trait)
                init.size = 7, #recruit size?
                establishment.logit = logit(0.67),
                y1surv.logit = logit(0.45),
                rel.growth = 1, #rate of accumulation of new dry mass per unit of existing dry mass
                flowering.logit = logit(0.25),
                fruitPerPlant = 30)

SP$addTraitA(nQtlPerChr = 10, 
              mean = traitMeans, 
              var = (traitMeans*0.25)^2, #genetic CV = 0.25 (?)
              gamma = TRUE, #a lot of small effect loci and a few large effect loci 
              shape = 1, #shape parameter for gamma distribution 
              name=names(traitMeans))
#addTraitAG - Randomly assigns eligible QTLs for one or more additive GxE traits. If simulating more than one trait, all traits will be pleiotropic with correlated effects.
#addTraitA - Randomly assigns eligible QTLs for one or more additive traits. If simulating more than one trait, all traits will be pleiotropic with correlated additive effects. (we will keep it simple for now)
##nQtlPerChr = number of QTLs per chromosome. Can be a single value or nChr values.
##mean = a vector of desired mean genetic values for one or more traits
##var = a vector of desired genetic variances for one or more traits
##varGxE = a vector of total genotype-by-environment variances for the traits
##name = optional name for trait(s)

SP$setVarE(h2=rep(0.5, length(traitMeans))) #set heritability to 0.5 
#Error in SP$setVarE(h2 = rep(0.5, length(traitMeans))) : all(private$.varG > 0) is not TRUE -  Y1 SURV = logit(0.5)
```

## Create Initial Population

```{r}
pop = newPop(founderPop, simParam = SP)
pop1 <- setPheno(pop)
```

```{r}
pheno <- pop1@pheno %>% as_tibble()
dim(pheno)
head(pheno)
```

### Convert pheno logits to probabilities and phenotypes
```{r}
pheno <- pheno %>%
  mutate(across(ends_with(".logit"), .fns = inv_logit, 
         .names = "{.col}.prob")) %>%
  rename_with(.fn = \(n) str_replace(n, "\\.logit\\.prob", "\\.prob")) %>%
  mutate(germinated = rbinom(n(), size=1, prob=germination.prob),
         established = ifelse(germinated, rbinom(n(), size=1, prob=establishment.prob), NA),
         init.size,
         rel.growth = ifelse(established, rel.growth, NA),
         y1surv = ifelse(established, rbinom(n(), size =1, prob=y1surv.prob), NA),
         flowered = ifelse(established, rbinom(n(), size = 1, prob = flowering.prob), NA),
         fruitPerPlant = ifelse(flowered, fruitPerPlant, NA))
```

```{r}
pheno %>%
  pivot_longer(everything()) %>%
  ggplot(aes(x=value)) +
  geom_histogram() +
  facet_wrap(~name, scales="free")
```

## Try to get weekly growth data 
If we know the initial size and the growth rate then we can get weekly size data, right?
RGR = (ln(S2)-ln(S1)) / (t2-t1))
t2-t1 = 1 week
RGR = ln(S2)-ln(S1)
RGR+ln(S1)=ln(S2)
```{r}
#weekly size function 
weekly.size <- function(s1, RGR, na.rm=FALSE) {
  if_else(s1==0, 0, 
          exp(RGR + log(s1)))
}

#weekly.size(7, 1) #function works 

pheno_weekly_test <- pheno %>% 
  filter(init.size>=0) %>%  #pheno has negatives for init.size - this is a temporary fix until I can figure out how to set the min as 0 
  mutate(size.week2=weekly.size(init.size, rel.growth, na.rm = TRUE))

#next step would be to automate this to generate the desired number of weeks 
##would probably want similar structure to "wl2_all_2023_timeprep" in "IntBio_WL2_2023.Rmd" script" - one row for each individual and each week
##maybe a for loop that loops around weeks to generates size t+1? and then bind_rows or something to combine all the output?
```

```{r}
#weekly size function 
weekly.size <- function(s1, RGR, na.rm=FALSE) {
  if_else(s1==0, 0, 
          exp(RGR + log(s1)))
}

pheno_weekly <- pheno %>% 
  mutate(Indiv_ID=row_number()) %>% 
  slice(rep(1:n(), each = 12)) %>% 
  group_by(Indiv_ID) %>% 
  mutate(week=row_number(),
         week_next=lead(week, order_by = week), 
         elapsed_days= week_next - week, #interval length in days,
         elapsed_weeks = as.integer(elapsed_days)/7) %>%  #interval length in weeks 
  filter(init.size>=0) %>%  #pheno has negatives for init.size - this is a temporary fix until I can figure out how to set the min as 0 
  mutate(size=if_else(week==1, init.size, NA)) %>% 
  mutate(size_prev=lag(size, order_by=week)) %>% 
  mutate(size=map(week, weekly.size, s1=size_prev, RGR=rel.growth, na.rm = TRUE))
#Caused by error in `.f()`:
#! unused argument (.x[[i]])
```

## Model growth with the Weibull model

```{r}
weibull <- function (t, alpha, beta, k, delta) {
  result <- alpha - (alpha - beta) * exp(-(k * t)^delta)
  return(result)
}

growth <- tibble(time = seq(0,100,.1)) %>%
  mutate(model1 = weibull(t = time,
                          alpha = 40, #Hmax = alpha -  Sets the asymptote (max growth)
                          beta = 0.5, #Hmin = beta - Changes the y intercept (the lower limit)
                          k = 0.01, #k - Often described as the growth rate --\> related to the slope of the line
                          delta = 0.5)) #delta - Sets the inflection point - when you shift b/t concave to convex

growth %>%  ggplot(aes(x=time, y=model1)) +
  geom_line()
```