---
title: "Population_Sim_Trial"
author: "Brandie QC"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Population Simulation Attempt

Suggested parameters from Julin:

-   1000 F2s (or just 1000 individuals)

    -   Done with runMacs

-   100 loci per trait with some kind of (expotnetial?) decay in effect
    size

    -   Use nQtlPerChr??

-   Traits:

    -   Germination

    -   Initial size

    -   Establishment

    -   Growth rate

    -   Flowering probability

        -   Ideally would depend on size... - do this through
            correlation matrix?

    -   Fruit per plant

        -   Ideally would depend on size...

    -   others?

Rishav demoed AlphaRSim:
<https://cran.r-project.org/web/packages/AlphaSimR/index.html>

-   Rishav's demo is on his GitHub:
    <https://github.com/rishavray/IPM_practice/blob/main/IPM_genetic_simulation.Rmd>

## Libraries

```{r}
library(AlphaSimR)
library(tidyverse)
logit <- brms::logit_scaled
inv_logit <- brms::inv_logit_scaled
```

## Create Founder Population

```{r}
founderPop = runMacs(
  nInd = 1000, #number of individuals to simulate 
  nChr = 10, #number of chromosomes to simulate - may not be important to match to Streps' 14
  segSites = NULL, #number of segregating sites to keep per chromosome.
  species = "GENERIC" #The GENERIC history is meant to be a reasonable all-purpose choice. It runs quickly and models a population with an effective populations size that has gone through several historic bottlenecks.
)

#With 100000 segSites get: Error in runMacs(nInd = 1000, nChr = 10, segSites = 1e+05, species = "GENERIC") : MaCS did not return enough segSites, use segSites=NULL to return all sites generated by MaCS
```

## Define Genetic Architecture for Traits

```{r}
SP <- SimParam$new(founderPop)
#Starts the process of building a new simulation by creating a new SimParam object and assigning a founder population to the class.

traitMeans <- c(germination.logit = logit(0.33), #logit to scale it (from gene --> trait)
                #init.size = 7, 
                establishment.logit = logit(0.67),
                y1surv.logit = logit(0.45),
                rel.growth = 1, 
                flowering.logit = logit(0.25),
                fruitPerPlant = 30)

SP$addTraitA(nQtlPerChr = 10, 
              mean = traitMeans, 
              var = (traitMeans*0.25)^2, #genetic CV = 0.25 (?)
              gamma = TRUE, #a lot of small effect loci and a few large effect loci 
              shape = 1, #shape parameter for gamma distribution 
              name=names(traitMeans))
#addTraitAG - Randomly assigns eligible QTLs for one or more additive GxE traits. If simulating more than one trait, all traits will be pleiotropic with correlated effects.
#addTraitA - Randomly assigns eligible QTLs for one or more additive traits. If simulating more than one trait, all traits will be pleiotropic with correlated additive effects. (we will keep it simple for now)
##nQtlPerChr = number of QTLs per chromosome. Can be a single value or nChr values.
##mean = a vector of desired mean genetic values for one or more traits
##var = a vector of desired genetic variances for one or more traits
##varGxE = a vector of total genotype-by-environment variances for the traits
##name = optional name for trait(s)

SP$setVarE(h2=rep(0.5, length(traitMeans))) #set heritability to 0.5 
#Error in SP$setVarE(h2 = rep(0.5, length(traitMeans))) : all(private$.varG > 0) is not TRUE -  Y1 SURV = logit(0.5)
```

## Create Initial Population

```{r}
pop = newPop(founderPop, simParam = SP)
pop1 <- setPheno(pop)
```

```{r}
pheno <- pop1@pheno %>% as_tibble()
dim(pheno)
head(pheno)
```

### Convert pheno logits to probabilities and phenotypes
```{r}
pheno <- pheno %>%
  mutate(across(ends_with(".logit"), .fns = inv_logit, 
         .names = "{.col}.prob")) %>%
  rename_with(.fn = \(n) str_replace(n, "\\.logit\\.prob", "\\.prob")) %>%
  mutate(germinated = rbinom(n(), size=1, prob=germination.prob),
         established = ifelse(germinated, rbinom(n(), size=1, prob=establishment.prob), NA),
         rel.growth = ifelse(established, rel.growth, NA),
         y1surv = ifelse(established, rbinom(n(), size =1, prob=y1surv.prob), NA),
         flowered = ifelse(established, rbinom(n(), size = 1, prob = flowering.prob), NA),
         fruitPerPlant = ifelse(flowered, fruitPerPlant, NA))
```

```{r}
pheno %>%
  pivot_longer(everything()) %>%
  ggplot(aes(x=value)) +
  geom_histogram() +
  facet_wrap(~name, scales="free")
```


## Simulate Multiple Generations with Changing Environment

We would be using the setPheno function to set the phenotypic values
with the residual environmental variance.

```{r}
results = list()
climateTrend = seq(1, 1.5, length.out = 10) # Warming climate scenario

for (gen in 1:10) {
  # Current climate variable
  currentClimate = climateTrend[gen]
  
  # Calculate fitness based on genetic value and environment
  pop = setPheno(pop, varE = VarE*currentClimate, reps = 1, simParam = SP)
  fitness = pop@pheno[,1]
  
  # Selection - individuals with higher fitness have higher probability of reproducing
  selectedIndices = sample(1:pop@nInd, 50, prob = fitness/sum(fitness))
  selectedPop = pop[selectedIndices]
  
  # Random mating
  pop = randCross(selectedPop, nCrosses = 100, nProgeny = 1, simParam = SP)
  
  # Store results
  results[[gen]] = data.frame(
    generation = gen,
    meanGV = mean(pop@gv),
    varGV = var(pop@gv[,1]),
    climate = currentClimate,
    fitness = mean(fitness)
  )
}
```

I'm clearly doing something wrong.

## Visualize Results

```{r}
# Convert results to data frame
resultsDF = do.call(rbind, results)

ggplot(resultsDF, aes(x = generation)) +
  geom_line(aes(y = meanGV, color = "Mean Genetic Value")) +
  geom_line(aes(y = climate, color = "Climate")) +
  geom_line(aes(y = varGV, colour = "Trait Variance"))+
  geom_line(aes(y =fitness, colour = "Fitness"))+
  labs(title = "Adaptation to Changing Climate",
       x = "Generation",
       y = "Value") +
  #scale_color_manual(values = c("Mean Genetic Value" = "blue", "Climate" = "red")) +
  theme_doc()
```
